### PR TITLE
Issue 7110: Need a command to remove table keys from Admin CLI

### LINKED ISSUE #7110: Need a command to remove-keys from Admin CLI
**Is your feature request related to a problem? Please describe.**
Currently there is no way to remove table keys from Admin CLI.

**Describe the solution you'd like**
Add new RemoveTableSegmentKeysCommand to Admin CLI.

**Describe alternatives you've considered**
There is no command in CLI for table segment to remove a table key.

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by AnkurAnand11**
**Change log description**  
Add RemoveTableSegmentKeysCommand to Admin CLI.

**Purpose of the change**  
Closes #7110 

**What the code does**  
Adds new RemoveTableSegmentKeysCommand (to remove key from segment) to Admin CLI.

**How to verify it**  
Execute the new admin command -` table-segment remove-key <qualified-table-segment-name> <key>`
Sample : `table-segment remove-key _system/_tables/testScope/streamsInScope.#.c09fcdf2-94fd-4b87-8e46-8f3aa6189237 testStream`
Response -> `RemoveTableKey: testStream removed successfully from _system/_tables/testScope/streamsInScope.#.c09fcdf2-94fd-4b87-8e46-8f3aa6189237`

### CODE REVIEW

**Comment by AnkurAnand11** [comment]
Ran development system test 2 successfully - build no. 2433

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeysCommand.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import io.pravega.client.tables.impl.HashTableIteratorItem;
+import io.pravega.client.tables.impl.TableSegmentKey;
+import io.pravega.shared.protocol.netty.PravegaNodeUri;
+import lombok.Cleanup;
+import org.apache.curator.framework.CuratorFramework;
+
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+public class RemoveTableSegmentKeysCommand extends TableSegmentCommand {
```
Comment (shshashwat) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeysCommand.java:30]: `javadoc` missing from `public` class

**Review by shshashwat — CHANGES_REQUESTED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeysCommand.java (hunk 1 above)
Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeysCommand.java]: Added javadoc.

**Review by tkaitchuck — APPROVED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+        @Cleanup
+        CuratorFramework zkClient = createZKClient();
+        @Cleanup
+        AdminSegmentHelper adminSegmentHelper = instantiateAdminSegmentHelper(zkClient);
+        CompletableFuture<HashTableIteratorItem<TableSegmentKey>> getKeysReply = adminSegmentHelper.readTableKeys(fullyQualifiedTableSegmentName,
+                new PravegaNodeUri(segmentStoreHost, getServiceConfig().getAdminGatewayPort()), Integer.MAX_VALUE,
+                HashTableIteratorItem.State.EMPTY, super.authHelper.retrieveMasterToken(), 0L);
+
+        TableSegmentKey keysList = getKeysReply.join().getItems()
```
Comment (sachin-j-joshi) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: It is not clear to me why we need to enumerate the keys first to remove them.
Why can't we remove key directly?
Will remove throw exception if key doesn't exist?

Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: No, it does not throw any exception if the key doesn't exist but just for the user to get a correct information that the key user is trying to remove is correct or not.

Comment (sachin-j-joshi) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: Please double check that no exception is thrown.

Unblocking for now

Comment (Bhupender-Y) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: What Sachin said make sense. I think if key doesn't exist then user will get DataNotFoundException.

File: controller/src/main/java/io/pravega/controller/server/SegmentHelper.java [hunk 3]
```diff
@@ -546,13 +546,22 @@ public CompletableFuture<Void> removeTableKeys(final String tableName,
...
                                                    final long clientRequestId) {
+
+        return removeTableKeys(tableName, ModelHelper.encode(uri), keys, delegationToken, clientRequestId);
+    }
+
+    public CompletableFuture<Void> removeTableKeys(final String tableName,
+                                                   PravegaNodeUri uri,
+                                                   final List<TableSegmentKey> keys,
+                                                   String delegationToken,
+                                                   final long clientRequestId) {
```
Comment (Bhupender-Y) [inline controller/src/main/java/io/pravega/controller/server/SegmentHelper.java:557]: Why we need to change these methods, can't we use the existing one?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java (hunk 2 above)
Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: It does not throw any exception if the key doesn't exist.

File: controller/src/main/java/io/pravega/controller/server/SegmentHelper.java (hunk 3 above)
Comment (AnkurAnand11) [reply controller/src/main/java/io/pravega/controller/server/SegmentHelper.java]: Updated to use the existing method.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                .stream()
+                .filter(tableSegmentKey -> getCommandArgs().getState().getKeySerializer().deserialize(getByteBuffer(tableSegmentKey.getKey())).equals(key))
+                .findFirst().orElse(null);
+
+        if (keysList != null) {
+            CompletableFuture<Void> reply = adminSegmentHelper.removeTableKeys(fullyQualifiedTableSegmentName, new PravegaNodeUri(segmentStoreHost, getServiceConfig().getAdminGatewayPort()),
+                    Collections.singletonList(TableSegmentKey.unversioned(key.getBytes(Charsets.UTF_8))), super.authHelper.retrieveMasterToken(), 0L);
+            reply.join();
```
[the remark this answers is not in the document]
Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java]: Moved the code inside try/catch block.

**Review by anishakj — APPROVED** [review]
LGTM

**Review by shshashwat — APPROVED** [review]

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        CuratorFramework zkClient = createZKClient();
+        @Cleanup
+        AdminSegmentHelper adminSegmentHelper = instantiateAdminSegmentHelper(zkClient);
+        try {
+            CompletableFuture<HashTableIteratorItem<TableSegmentKey>> getKeysReply = adminSegmentHelper.readTableKeys(fullyQualifiedTableSegmentName, Integer.MAX_VALUE,
+                    HashTableIteratorItem.State.EMPTY, super.authHelper.retrieveMasterToken(), 0L);
+
+            TableSegmentKey keysList = getKeysReply.join().getItems()
+                    .stream()
+                    .filter(tableSegmentKey -> getCommandArgs().getState().getKeySerializer().deserialize(getByteBuffer(tableSegmentKey.getKey())).equals(key))
```
Comment (abhinb) [inline cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java:61]: I think you dont need to get the keyserializer from AdminCommand state here. Doing so would mean we are putting a dependancy on serializers like `container_meta or slts and they are for specific tables.  You can just choose to convert the bytebuf to string instead and remove the pre-requisite to set serializers right?

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+            } else {
+                output("RemoveTableKey failed: %s does not exist", key);
+            }
+        } catch (Exception ex) {
+            output("RemoveTableKey failed: %s", ex.getMessage());
+        }
+    }
+
+    public static CommandDescriptor descriptor() {
+        return new CommandDescriptor(COMPONENT, "remove-key", "Removes table segment key. Use the command \"table-segment set-serializer <serializer-name>\" to use the appropriate serializer before using this command.",
```
[the remark this answers is not in the document]
Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java]: Removed setting the serializer.

File: cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java (hunk 5 above)
Comment (AnkurAnand11) [reply cli/admin/src/main/java/io/pravega/cli/admin/segmentstore/tableSegment/RemoveTableSegmentKeyCommand.java]: Updated.

**Review by abhinb — APPROVED** [review]

**Pull request merged**